File: day13.c
#include <stdio.h>
#include <stdint.h>

int64_t cost(int a, int b, int c, int d, int64_t e, int64_t f){
    int64_t N1=e*d-f*c,N2=a*f-b*e,D=a*d-b*c;
    if(N1%D==0 && N2%D==0){return 3*N1/D+N2/D;}else{return 0;}
}

int main(void){
    int64_t part1=0,part2=0;
    FILE *file=fopen("data/input13.txt","r");
    int a,b,c,d,e,f;
    int64_t B=10000000000000LL;
    while(fscanf(file,"Button A: X+%d, Y+%d\n",&a,&b)==2){
        fscanf(file,"Button B: X+%d, Y+%d\n",&c,&d);
        fscanf(file,"Prize: X=%d, Y=%d\n",&e,&f);
        part1+=cost(a,b,c,d,e,f);
        part2+=cost(a,b,c,d,e+B,f+B);
    }
    printf("%lld %lld\n",part1,part2);
    return 0;
}

File: day13.py
import re
with open("data/input13.txt") as f: L=[int(x) for x in re.findall("\d+",f.read())]
machines=[]
for i in range(0,len(L),6):
    machines.append({'A':(L[i],L[i+1]),'B':(L[i+2],L[i+3]),'prize':(L[i+4],L[i+5])})

def det(u,v): return u[0]*v[1]-u[1]*v[0]
def coords(v,u1,u2):
    D,N1,N2=det(u1,u2),det(v,u2),det(u1,v)
    if N1%D==0 and N2%D==0: return (N1//D, N2//D)
    return None
def cost(v):
    if v:return 3*v[0]+v[1]
    return 0
def tr(v): return (v[0]+10**13,v[1]+10**13)

part1=sum(cost(coords(m['prize'],m['A'],m['B'])) for m in machines)
part2=sum(cost(coords(tr(m['prize']),m['A'],m['B'])) for m in machines)
print(part1,part2)